Improved FAQ documentation some more (#429)
reactive-firewall-org/multicast: docs/USAGE.md

## Basic Library Usage

> The API is in a continuous state of improvement, so version pinning is recommended for now.

Author: :thinking: good idea; consider this:
```suggestion
> [!NOTE]
> The API is in a continuous state of improvement, so version pinning is recommended for now.
```

### Listening to Multicast

> The preferred way to receive multicast messages is to allow the `multicast.hear.McastServer`
> template class handle the ephemeral multicast receivers as a kind of multicast server, allowing
> a handler to focus on processing the data.

Author: :thinking: that does sound better when read aloud; consider this:
```suggestion
The preferred way to receive multicast messages is to allow the `multicast.hear.McastServer`
template class to handle the ephemeral multicast receivers as a kind of multicast server, allowing
a handler to focus on processing the data.
```

a handler to focus on processing the data.

Here is an example of multicast socket server usage (circa v2.1)

```python3
# imports
import multicast
from multiprocessing import Process
import random  # for random port

# Multicast group address and port
MCAST_GRP = "224.0.0.1"  # Replace with your multicast group address (use IPv4 dotted notation)

> MCAST_PORT = int(random.SystemRandom().randint(49152, 65535))  # Replace with your multicast port

Author: 🙅🏻 The current implementation is fine. The reader is supposed to replace the whole value anyway, the redundant `int()` is stylistic exception for readability in this example, because the random function is slightly less obvious (albeit already reasonably clear). Does that make sense? or do you still find fault with my reasoning?

#    Windows users must use option 1 or 2 for now.
# This address is per socket (e.g., can be chosen per socket even if on a single interface)
# HOST_BIND_IP = "0.0.0.0"

# Options for multicast listener
listener_options = {
    "is_daemon": True,  # bool: enable daemon mode
    "port": MCAST_PORT,  # int: UDP port for multicast
    "group": MCAST_GRP  # str: multicast group address (use IPv4 dotted notation)
}

# Create a multicast listener
listener = multicast.hear.McastHEAR()

# create a separate process for the listener
p = Process(
    target=listener,
    name="HEAR", kwargs=listener_options
)
p.daemon = listener_options["is_daemon"]
p.start()

# ... use CTL+C (or signal 2) to shutdown the server 'p'
```

> [!IMPORTANT]
> The above example probably will return with nothing outside a handler function in a loop,
> unless you enable the default logging beforehand with:
>
> ```python3
> # setup console logging as example
> import logging
> multicast_logging_sink = logging.getLogger()
> multicast_logging_sink.setLevel(logging.INFO)  # increase default logging from multicast module
> handler = logging.StreamHandler()  # example trivial log handler

> > multicast_logging_sink.addHandler(handler)

Author: Actually, the issue here is twofold: first, log-level, and second, the default module handler is a null log handler so an explicit choice is required to add a log handler for the module log records,

🙅🏻 The current implementation is better.

> > multicast_logging_sink.addHandler(handler)

Author: @coderabbitai, for posterity please draft a FAQ style question and answer about this discussion. I look forward to reviewing your draft.

# The sender will default to letting the system choose
# HOST_BIND_IP = "0.0.0.0"

# Options for multicast sender
sender_options = {
    "port": MCAST_PORT,  # int: UDP port for multicast
    "group": MCAST_GRP,  # str: multicast group address (use IPv4 dotted notation)
    "data": "Default listener only knows: STOP"  # str: message content to try to _transmit_
}

# Create an ephemeral multicast sender
sender = multicast.send.McastSAY()

# create a separate process for the sender
p = Process(
    target=sender,
    name="SAY", kwargs=sender_options
)
p.daemon = False  # sender should not be a daemon

try:
    p.start()
except Exception as baton:
    p.join()  # good practice to handle clean up
    raise RuntimeError("multicast seems to have failed.") from baton  # re-raise
finally:
    # clean up some stuff
    if p:
        p.join() # if not already handled don't forget to join the process and other overhead
    # hint: if you use a loop and need to know the exit code
    didWork = (p is not None and int(p.exitcode) <= int(0))  # e.g. check for success

```

## Advanced Library Usage

### Custom handlers

> > [!TIP]
> > The API for custom handlers currently requires implementing a sub-class
> > `multicast.hear.HearUDPHandler` and handling the listener's `multicast.hear.McastServer` server

Author: so like this:
```suggestion
> [!TIP]
> The API for custom handlers currently requires implementing a subclass
> `multicast.hear.HearUDPHandler` and handling the listener's `multicast.hear.McastServer` server
```

## Basic Library Usage

> [!NOTE]
> The API is in a continuous state of improvement, so version pinning is recommended for now.

### Listening to Multicast

The preferred way to receive multicast messages is to allow the `multicast.hear.McastServer`
template class to handle the ephemeral multicast receivers as a kind of multicast server, allowing
a handler to focus on processing the data.

Here is an example of multicast socket server usage (circa v2.1)

```python3

> # imports

Author: 🙅🏻 The current implementation is fine.

try:
    p.start()
except Exception as baton:
    p.join()  # good practice to handle clean up
    raise RuntimeError("multicast seems to have failed.") from baton  # re-raise
finally:
    # clean up some stuff
    if p:
        p.join() # if not already handled don't forget to join the process and other overhead
    # hint: if you use a loop and need to know the exit code
    didWork = (p is not None and p.exitcode <= 0)  # e.g. check for success

```

## Advanced Library Usage

### Custom handlers

> [!TIP]
> The API for custom handlers currently requires implementing a subclass
> `multicast.hear.HearUDPHandler` and handling the listener's `multicast.hear.McastServer` server
> directly with something like this:
>
> ```python3
> from multicast.hear import McastServer, HearUDPHandler
> with McastServer((MCAST_GRP, MCAST_PORT), HearUDPHandler) as server:
>     server_initialized = True
>     server.serve_forever()  # ... use CTL+C (or signal 2) to shutdown the server
> ```
>
> This is essentially what the
> [default listener does under-the-hood](https://github.com/reactive-firewall/multicast/blob/v2.0.9a5/multicast/hear.py#L830C4-L832C27)

> > automatically.

Author: 🙅🏻 The current implementation is fine for now as v2.1 is the next release, and may become harder to maintain if pinned to a branch like "stable" because even a small addition will cause lines to move, however by pinning the current implementation the link does not need to be updated after the next release nor maintained for every addition, only for superseding/functional changes to the relevant code.
